---
title: "Assignment1_Redo"
author: "Shrija Srinivasan"
date: "22/10/2019"
output: pdf_document
---

Question Addressed: Attached File is a count table from the software "Salmon". There are 3 samples in "WildType" group. Find out the differentially expressed transcripts between mutant and wildtype. Use DESeq2 with package "tximport".

As transcripts could be many and would arise from one gene, it would be nice to associate it to a gene, so we could have a count table easily. After creating a vector to the quantification files, the samples have been read and analysed accordingly to associate the transcripts with gene-ID for summary at gene-level. We try to load data for tximport. To remember that there are two kinds of data available. One is the extdata having the annotation files and the other are the Test data(which we have at the moment) and we integrate them into R studio. The expression matrix will be loaded into R using the read.table function. The count table will contain the counts of each gene(row) and the each sample(column)

##Phenotypic Data

The dataset contains the wildtype(WT) and mutant(M) strains  which work with a certain number of biological replicates each genotype. The required information on the phenotype is enclosed in a phenotypic file. We try to load the data and view it.

```{r implementation to demonstrate tximport and readr package}
##Creating a directory to load all the datas
dir<- file.path("/home/shrija/Documents/UjjwalNeogiLab")
list.files(dir)

##Assigning the names of the sf files
valuename<-c("quant_mut1.sf","quant_mut2.sf","quant_mut3.sf","quant_wt1.sf","quant_wt2.sf","quant_wt3.sf")
files<-file.path(dir, "Test_Data", valuename)
names(files) <- c(paste0("M_",1:3), paste0("WT-",1:3))
files

#Reading readr and tximport
library(readr)
library(tximport)

##Phenotypic Data
ExperimentalDesign<-tximport(files,type = "salmon",txOut =TRUE)
head(ExperimentalDesign)
tail(ExperimentalDesign)

##Count the number of sample/abundance in each class
table(ExperimentalDesign$abundance)

## Dimensions
samplesize<-ncol(ExperimentalDesign$counts)
print(samplesize)

samplegenes<-nrow(ExperimentalDesign$counts)
print(samplegenes)

dim(ExperimentalDesign$counts)

##Summary of the Table
head(summary(ExperimentalDesign$counts))
#statsofeachsample<-data.frame(t(do.call(cbind,lapply(ExperimentalDesign$counts, summary))))
#head(statsofeachsample)
#tail(statsofeachsample)

##Histogram plot of the count
hist(as.matrix(ExperimentalDesign$counts),col="purple",border = "White",breaks =10000, xlim=c(0,2000),main="Counts per gene", xlab="Counts(truncated axis)",ylab="Number of Genes",las=1,cex.axis=0.7)

##In order to avoid issues with log(0)
epsilon<-1
hist(as.matrix(log2(ExperimentalDesign$counts + epsilon)),breaks=50,col="purple",border = "white",main = "log2 transformed counts per gene",xlab="log2(Counts+1)",ylab="Number of genes", las=1, cex.axis=0.7)
```
##Interpretation
The top histogram is not very informative due to the presence of high count values towards the x-axis(due to its large x-axis scale).
The logrithmic transformation improves the readability of the maps. We added the count value to 1 to avoid the problems with log-transformation of zero-counts.(which results in negative infinity)


##Boxplot Interpretation
To have a more clear insight into the distribution of the sample, we try to offer using a boxplot.

##Eliminating Undetected Genes
There are a set of genes that remain undetected and has to be removed. This is because only a few express well to be detected while some tend to weakly express that they are hardly detected in the sample.
As a result the sample may contain only null values. or zero values. Certain questions need to be addressed like:
What is the percentage of the gene having null count per sample. Draw barplot
Some genes were not detected in the sample. Count the required number, and delete from the count table.

```{r boxplots}
boxplot(log2(ExperimentalDesign$counts +epsilon),pch=".",col="purple", horizontal=TRUE, cex.axis=0.5,las=1,ylab="Samples",xlab="log2(counts+1)")

##Eliminating undetected genes and drawing required barplot
tablewithnull<-apply(ExperimentalDesign$counts, 2, function(x) 100*mean(x==0))
print(tablewithnull)

barplot(tablewithnull,main = "Percentage of null counts per sample",horiz = TRUE, cex.names = 0.5, las=1, col="purple",ylab ="Samples",xlab = "% of Null counts")

#Differential Expression Analysis
library(DESeq2)

#Creating the required dataset for the DESeq2
sampleTable <- data.frame(condition = factor(rep(c("Mutant", "WildType"), each = 3)))
rownames(sampleTable) <- colnames(ExperimentalDesign$counts)
dds <- DESeqDataSetFromTximport(ExperimentalDesign, sampleTable, ~condition)
print(dds)

#What kind of object is it?
is(dds)
isS4(dds)

#What does it contain?
slotNames(dds)
```


##Scaling Factor
It is very crucial to have a required scaling factor or a normalisation method for estimating the samples and genes correctly.

```{r Implementation}
##Let's implement this function
cds<-ExperimentalDesign$counts

ddsnormal <-  estimateSizeFactors(dds)
sizeFactors(ddsnormal)

## Computing mean and variance
normcounts <- counts(ddsnormal, normalized=TRUE)
meancounts <- rowMeans(normcounts)
variance.counts <- apply(normcounts, 1, var)

## sum(mean.counts==0) # Number of completely undetected genes

normcountsstats <- data.frame(
  min=apply(normcounts, 2, min),
  mean=apply(normcounts, 2, mean),
  median=apply(normcounts, 2, median),
  max=apply(normcounts, 2, max),
  zeros=apply(normcounts==0, 2, sum),
  percentzeros=100*apply(normcounts==0, 2, sum)/nrow(normcounts),
  perc05=apply(normcounts, 2, quantile, 0.05),
  perc10=apply(normcounts, 2, quantile, 0.10),
  perc90=apply(normcounts, 2, quantile, 0.90),
  perc95=apply(normcounts, 2, quantile, 0.95)
)
variancecounts <- apply(normcounts, 1, var)
print(normcountsstats)

##Mean and Variance Relationship
meanvarcol <- densCols(x=log2(meancounts), y=log2(variancecounts))
{plot(x=log2(meancounts), y=log2(variancecounts), pch=16, cex=0.5, 
     col=meanvarcol, main="Mean-variance relationship",
     xlab="Mean log2(normalized counts) per gene",
     ylab="Variance of log2(normalized counts)",
     panel.first = grid())
abline(a=0, b=1, col="brown")}

##Modelling readcounts through a negative binomial
ddsdispersion<- estimateDispersions(ddsnormal)


##A diagnostic plot which shows the mean of normalised counts(of x-axis) and dispersion estimate of each genes
plotDispEsts(ddsdispersion)

##Performing Differential Expression Call
alpha <- 0.0001
waldtest <- nbinomWaldTest(ddsdispersion)
resDESeq2 <- results(waldtest, alpha=alpha, pAdjustMethod="BH")
is(resDESeq2) # a data.frame
head(resDESeq2)

#The colnames containing the word 'padj' contains the FDR values obtained from Benjamini-Hochberg procedure
colnames(resDESeq2)

#Adjusting based on the decreasing p-value
resDESeq2 <- resDESeq2[order(resDESeq2$padj),]
head(resDESeq2)

## Draw an histogram of the p-values
hist(resDESeq2$padj, breaks=20, col="grey", main="DESeq2 p-value distribution", xlab="DESeq2 P-value", ylab="Number of genes")

##Volcano Plot
alpha <- 0.01 # Threshold on the adjusted p-value
cols <- densCols(resDESeq2$log2FoldChange, -log10(resDESeq2$pvalue))
{plot(resDESeq2$log2FoldChange, -log10(resDESeq2$padj), col=cols, panel.first=grid(),
     main="Volcano plot", xlab="Effect size: log2(fold-change)", ylab="-log10(adjusted p-value)",
     pch=20, cex=0.6) 
abline(v=0) 
abline(v=c(-1,1), col="brown")
abline(h=-log10(alpha), col="brown")

geneselected <- abs(resDESeq2$log2FoldChange) > 2 & resDESeq2$padj < alpha 
text(resDESeq2$log2FoldChange[geneselected],
     -log10(resDESeq2$padj)[geneselected],
     lab=rownames(resDESeq2)[geneselected ], cex=0.4)}

##To find out the most differentially expressed gene
genemostsignificant <- rownames(resDESeq2)[1]
genemostdiffvalue <- counts(ddsnormal, normalized=T)[genemostsignificant,]
barplot(genemostdiffvalue, main=genemostsignificant, las=2, cex.names=0.5)

##Looking at hte results with MA plot
#The genes that are given with below 1% are added here accordingly
{plotMA(resDESeq2,colNonSig="blue")
abline(h=c(-1:1), col="red")}
```

##Hierarchical Clustering
To ensure that the selected genes distinguish well between "Wildtype" and "Mutant", we will perform a clustering using heatplot from ggplot library.

```{r hierarchical clustering}
genekept<- rownames(resDESeq2)[resDESeq2$padj <= alpha & !is.na(resDESeq2$padj)]

## We retrieve the normalized counts for gene of interest
counttablekept <- log2(cds + epsilon)[genekept, ]
dim(counttablekept)

## Install the gplots library if needed then load it
if(!require("gplots")){
  install.packages("gplots")
}
library("gplots")

## Perform the hierarchical clustering with a distance based on Pearson-correlation coefficient and average linkage clustering as agglomeration criteria
heatmap.2(as.matrix(counttablekept), 
          scale="row", 
          hclust=function(x) hclust(x,method="average"), 
          distfun=function(x) as.dist((1-cor(t(x)))/2), 
          trace="none", 
          density="none", 
          labRow="",
          cexCol=0.7)

install.packages("gProfileR")
library(gProfileR)

resDESeq2.df <- na.omit(data.frame(resDESeq2))
upregulatedtranscripts <- rownames(resDESeq2.df)[resDESeq2.df$log2FoldChange >= 2 &  resDESeq2.df$padj < alpha]
head(upregulatedtranscripts)
names(terminduced)

resDESeq2.df <- na.omit(data.frame(resDESeq2))
downregulatedtranscripts <- rownames(resDESeq2.df)[resDESeq2.df$log2FoldChange <= -2 &  resDESeq2.df$padj < alpha]
head(downregulatedtranscripts)
```
